File: steps/typescript/delete-pet.step.ts
// steps/typescript/delete-pet.step.ts
import { ApiRouteConfig, Handlers } from 'motia';
import { z } from 'zod';
import { TSStore } from './ts-store';

// Define path parameter schema
const pathParamsSchema = z.object({
  id: z.string().min(1, 'Pet ID is required')
});

export const config: ApiRouteConfig = {
  type: 'api',
  name: 'TsDeletePet',
  path: '/ts/pets/:id',
  method: 'DELETE',
  emits: [],
  flows: ['TsPetManagement']
};

export const handler: Handlers['TsDeletePet'] = async (req, { logger }) => {
  try {
    // Validate path parameters
    const { id } = pathParamsSchema.parse(req.pathParams);

    const deletedPet = TSStore.softDelete(id);

    if (!deletedPet) {
      return { status: 404, body: { message: 'Pet not found' } };
    }

    if (logger) {
      logger.info('🗑️ Pet soft deleted', {
        petId: deletedPet.id,
        name: deletedPet.name,
        purgeAt: new Date(deletedPet.purgeAt!).toISOString()
      });
    }

    return {
      status: 202,
      body: {
        message: 'Pet scheduled for deletion',
        petId: deletedPet.id,
        purgeAt: deletedPet.purgeAt
      }
    };

  } catch (error) {
    if (error instanceof z.ZodError) {
      return {
        status: 400,
        body: {
          message: 'Validation error',
          errors: error.errors
        }
      };
    }

    return {
      status: 500,
      body: { message: 'Internal server error' }
    };
  }
};

File: steps/python/delete_pet_step.py
# steps/python/delete_pet.step.py
config = { "type":"api", "name":"PyDeletePet", "path":"/py/pets/:id", "method":"DELETE", "emits": [], "flows": ["PyPetManagement"] }

async def handler(req, ctx=None):
    logger = getattr(ctx, 'logger', None) if ctx else None
    emit = getattr(ctx, 'emit', None) if ctx else None
    
    try:
        import sys
        import os
        import time
        sys.path.append(os.path.join(os.path.dirname(__file__), '..'))
        from services import pet_store
    except ImportError:
        return {"status": 500, "body": {"message": "Import error"}}
    
    pet_id = req.get("pathParams", {}).get("id")
    deleted_pet = pet_store.soft_delete(pet_id)
    
    if not deleted_pet:
        return {"status": 404, "body": {"message": "Not found"}}

    if logger:
        logger.info('🗑️ Pet soft deleted', {
            'petId': deleted_pet['id'],
            'name': deleted_pet['name'],
            'purgeAt': time.strftime('%Y-%m-%dT%H:%M:%S', time.gmtime(deleted_pet['purgeAt'] / 1000))
        })

    if emit:
        await emit({
            'topic': 'py.pet.soft.deleted',
            'data': {
                'petId': deleted_pet['id'],
                'name': deleted_pet['name'],
                'purgeAt': deleted_pet['purgeAt']
            }
        })

    return {
        "status": 202,
        "body": {
            "message": "Pet scheduled for deletion",
            "petId": deleted_pet['id'],
            "purgeAt": deleted_pet['purgeAt']
        }
    }